src/kle/forth.396

```diff
@@ -0,0 +1,800 @@
+	      (setq rtn
+		    (cond ((not (zerop (pop parameter-stack)))
+			   (pop parameter-stack))
+			  (t nil)))
+	      (store (memory H)
+		     (- (memory H) 1))
+	      (return rtn)))
+
+
+		;SHOW finds a word in the dictionary and shows its entry
+
+(defun show macro (n)
+       `(let ((loc (find ,(cadr n))))
+	     (cond ((null loc)
+		    (error '|  show: word not found| ,(cadr n)))
+		   (t (dp 'memory loc (+ loc showsize))))))
+
+		;LD lists the entries in the dictionary
+		;memory location at a time (unreadable)
+(defun ld ()
+       (do ((a (memory context) (memory (1+ a)))
+	    (i 1 (1+ i)))	   
+	   ((= 0 a) nil)
+	   (progn (print i)
+		  (princ '| - |)
+		  (princ a)
+		  (princ '| -> |)
+		  (princ (memory a)))))
+
+
+
+
+(defun er ()
+       (errprint nil)		;print the last error message
+       (baktrace)		;print out function stack
+       nil)
+
+(defun make macro (n)
+       `(store (memory ,(cadr n)) ,(caddr n)))
+
+(defun s macro (n)
+       `(memory ,(cadr n)))
+
+(defun s@ macro (n)
+       `(locate ,(cadr n)))
+
+(defun mark ()
+       (setq oldh (s h))
+       (setq oldcontext (s context))
+       (setq oldparameter-stack parameter-stack)
+       (stat) )
+
+(defun remark ()
+       (store (memory h) oldh)
+       (store (memory context) oldcontext)
+       (setq parameter-stack oldparameter-stack)
+       (stat) )
+
+
+(defun stat ()
+       (print 'H-->) (princ (s h))
+       (print '@H->) (princ (s@ h))
+       (print 'CONTEXT-->) (princ (s context))
+       (print '@CONTEXT->) (princ (s@ context))
+       (print 'parameter-stack->) (princ parameter-stack))
+
+(defun write macro (n)		;(write	<file> <s-exp>)
+       `((lambda (outfiles ^r ^w)
+		 ,(caddr n)
+		 (close (car outfiles)))
+	 (list (open ,(cadr n) 'out)) t t))
+
+(defun document macro (n)
+	`(write ,(cadr n)
+		 (progn (stat)
+			(terpri)
+			(pd))))
+
+(defmacro save (file)
+	  `(filewrite ',file
+		      (print `(fillarray 'memory
+					 ',(listarray 'memory)))))
+
+;-------------------------------------------------------------------------------
+
+	;;; RANDOM  DECLARATIONS
+
+(setq base 10.		;And foo said let thy input and output be in decimal.
+      ibase 10.		;And in the beginning, bar saw that it was good.
+      *nopoint t)	;And yea verily thou shalt not print decimal points.
+
+
+(setq FALSE 0)		;TRUE DEFINED NON-ZERO
+
+
+
+		;; Additional support routines
+
+				;Turn translation off i.e. distinguish between
+				;upper and lower case.
+
+(DEFUN TRANS-OFF ()
+       (MAPC '(LAMBDA (X) (SETSYNTAX X NIL X))
+	     (EXPLODEC '|abcdefghijklmnopqrstuvwxyz|)))
+
+
+				;Turn translation on i.e. do not distinguish
+				;between upper and lower case [default] 
+
+(DEFUN TRANS-ON ()
+       (MAPC '(LAMBDA (X Y ) (SETSYNTAX X NIL Y))
+	     (EXPLODEC '|abcdefghijklmnopqrstuvwxyz|)
+	     (EXPLODEC '|ABCDEFGHIJKLMNOPQRSTUVWXYZ|)))
+
+
+
+		;Load the file to bootstrap the forth dictionary.
+
+(let ((fet (status features)))
+     (cond ((member 'mc fet)
```

Comment:
Should this just do `(load '|dsk:kle;fordic >|)` unconditionally, rather than checking which machine it's on?